spring-integration-file/src/main/java/org/springframework/integration/file/aggregator/FileMarkerReleaseStrategy.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.file.aggregator;
+
+import java.util.Collection;
+
+import org.springframework.integration.aggregator.ReleaseStrategy;
+import org.springframework.integration.file.FileHeaders;
+import org.springframework.integration.file.splitter.FileSplitter;
+import org.springframework.integration.store.MessageGroup;
+import org.springframework.messaging.Message;
+
+/**
+ * A {@link ReleaseStrategy} which makes a decision based on the presence of
+ * {@link org.springframework.integration.file.splitter.FileSplitter.FileMarker.Mark#END}
+ * message in the group and its {@link org.springframework.integration.file.FileHeaders#LINE_COUNT} header.
+ *
+ * @author Artem Bilan
+ *
+ * @since 5.5
+ */
+public class FileMarkerReleaseStrategy implements ReleaseStrategy {
+
+	@Override public boolean canRelease(MessageGroup group) {
+		Collection<Message<?>> messages = group.getMessages();
+
+		return messages
+				.stream()
+				.filter((message) ->
+						FileSplitter.FileMarker.Mark.END.name()
+								.equals(message.getHeaders().get(FileHeaders.MARKER)))
+				.findAny()
+				.map((message) -> message.getHeaders().get(FileHeaders.LINE_COUNT, Long.class))
+				.map((lineCount) -> lineCount == messages.size() - 2)
+				.orElse(false);
```

Comment:
This won't scale well, especially with large files; it would be better to add `getLast()` to message group.

Reply:
Oh! I see your point. It looks like the `ReleaseStrategy` contract would be better as `canRelease(MessageGroup group, Message<?> currentMessage)` since we always trigger it whenever we got a new message. 
We can deprecate the current contract in favor of new and remove in the next `6.0` 😄 

But `getLast()` API in the `MessageGroup` won't hurt!

Reply:
Well, after looking into this closely, the `getLast()` doesn't feel OK. My test with the `filter()` and `ExecutorChannel` confirms that the last one is not always the `END` marker. So, I would say that `canRelease(MessageGroup group, Message<?> currentMessage)` makes sense for the main code flow, but `forceComplete(MessageGroup)` is still need to iterate the whole group, since there is no way to be sure that the last added to the group is exactly a `FileMarker` message.

At least some optimization is on its way! 😄

Reply:
Well, if they go async, all bets are off. If the last marker is processed before the size matches, we'll never release the group.

If we want to support that (async), we'd need another api on the group (e.g. `condition1Passed`, set to true if we ever received the last marker).

Also, the file lines would be re-assembled out of order.

Reply:
We may not care about an order of those items, especially when we convert them to some domain objects for further batch inserts into DB. The `applySequence` on the `FileSplitter` and `Resequencer` were always there for us to reassemble an original order.

I need to give the problem you have described more thoughts...

Have a good weekend!